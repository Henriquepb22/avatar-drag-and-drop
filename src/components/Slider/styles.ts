import styled, { css } from 'styled-components'

export const SliderContainer = styled.div`
    display: flex;
    flex-direction: column;
`

export const SliderLabel = styled.label`
    ${({ theme }) => css`
        font-size: ${theme.font.sizes.medium};
        line-height: 180%;
        letter-spacing: -0.02em;
        color: ${theme.colors.grayDark};
        margin-bottom: ${theme.spacings.xsmall};
    `}
`

export const Slider = styled.input`
    cursor: pointer;
    height: 0.2rem;
    width: 100%;
`
