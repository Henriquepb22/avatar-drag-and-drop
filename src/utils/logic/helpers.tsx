const isImageFile = (file: File) => {
    const fileType = file.type
    const baseMimeType = fileType.split('/')[0]

    if (baseMimeType !== 'image') return false
    return true
}

export const getFileDataUrl = async (file: File): Promise<string> => {
    const reader = new FileReader()

    return new Promise((resolve, reject) => {
        const isImageType = isImageFile(file)

        if (isImageType) {
            reader.readAsDataURL(file)
            reader.onload = () => {
                if (reader.result) {
                    resolve(reader.result.toString())
                }
            }
        } else {
            reader.abort()
            reject(new Error('File is not an recognized image type'))
        }
    })
}

export const cropImage = async (
    imageSrc: string,
    zoomLevel: number
): Promise<string> => {
    const canvas = document.createElement('canvas')
    const context = canvas.getContext('2d')
    const image = new Image()

    return new Promise((resolve, reject) => {
        image.src = imageSrc

        image.onload = () => {
            if (context) {
                const needCompression =
                    Math.max(canvas.width, canvas.height) > 1000

                canvas.width = needCompression ? image.width / 2 : image.width
                canvas.height = needCompression
                    ? image.height / 2
                    : image.height

                context.drawImage(image, 0, 0, canvas.width, canvas.height)
                context.globalCompositeOperation = 'destination-in'
                context.beginPath()
                context.arc(
                    canvas.width * 0.5,
                    canvas.height * 0.5,
                    (Math.max(canvas.width, canvas.height) * 0.5) /
                        (zoomLevel * 2),
                    0,
                    2 * Math.PI
                )
                context.fill()
                context.globalCompositeOperation = 'source-over'

                resolve(canvas.toDataURL('image/png', 1))
            }
        }

        image.onerror = () => {
            reject(new Error('Something bad happened...'))
        }
    })
}
