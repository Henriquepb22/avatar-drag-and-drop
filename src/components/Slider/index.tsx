import { useEffect, useState } from 'react'

import * as S from './styles'

type SliderProps = {
    onChange: (value: number) => void
}

const Slider = ({ onChange }: SliderProps) => {
    const [value, setValue] = useState(5)

    useEffect(() => {
        onChange(value)
    }, [value, onChange])

    return (
        <S.SliderContainer>
            <S.SliderLabel htmlFor="slider">Crop</S.SliderLabel>
            <S.Slider
                id="slider"
                type="range"
                value={value}
                onChange={(e) => setValue(e.currentTarget.valueAsNumber)}
                max={10}
            />
        </S.SliderContainer>
    )
}

export default Slider
