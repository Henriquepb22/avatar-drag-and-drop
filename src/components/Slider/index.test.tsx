import { screen, fireEvent } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'

import Slider from '.'

describe('<Slider />', () => {
    it('should render slider correctly', () => {
        const { container } = renderWithTheme(<Slider onChange={jest.fn} />)

        expect(screen.getByRole('slider')).toBeInTheDocument()
        expect(container).toMatchSnapshot()
    })

    it('should call onChange when value change', () => {
        const handleCrop = jest.fn()
        renderWithTheme(<Slider onChange={handleCrop} />)

        const slider = screen.getByRole('slider')
        fireEvent.change(slider, { target: { value: 1 } })

        expect(handleCrop).toHaveBeenCalledTimes(2)
    })
})
