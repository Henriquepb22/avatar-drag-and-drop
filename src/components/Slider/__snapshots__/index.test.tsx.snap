// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Slider /> should render slider correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  font-size: 1.6rem;
  line-height: 180%;
  -webkit-letter-spacing: -0.02em;
  -moz-letter-spacing: -0.02em;
  -ms-letter-spacing: -0.02em;
  letter-spacing: -0.02em;
  color: #677489;
  margin-bottom: 0.8rem;
}

.c2 {
  cursor: pointer;
  height: 0.2rem;
  width: 100%;
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1"
      for="slider"
    >
      Crop
    </label>
    <input
      class="c2"
      id="slider"
      max="10"
      type="range"
      value="5"
    />
  </div>
</div>
`;
